examples/llm/tests/test_moe.py

```diff
@@ -0,0 +1,183 @@
+# Copyright 2022 MosaicML Examples authors
+# SPDX-License-Identifier: Apache-2.0
+
+import os
+import warnings
+
+import pytest
+import torch
+from composer import Trainer
+from composer.core import Evaluator
+from composer.utils import dist, reproducibility
+from omegaconf import OmegaConf as om
+
+from examples.common.builders import (build_algorithm, build_callback,
+                                      build_dataloader, build_logger,
+                                      build_optimizer, build_scheduler)
+from examples.common.config_utils import log_config
+from examples.llm.src.model_registry import COMPOSER_MODEL_REGISTRY
+
+
+def gpt_tiny_moe_cfg(conf_path='yamls/mosaic_gpt/125m_moe.yaml'):
+    """Create gpt tiny moe cfg."""
+    with open(conf_path) as f:
+        test_cfg = om.load(f)
+    test_cfg.train_loader.dataset.split = 'train_small'
+
+    test_cfg.model.init_device = 'cpu'
+
+    bsize = 2
+    test_cfg.device_eval_batch_size = bsize
+    test_cfg.device_train_microbatch_size = bsize
+    test_cfg.global_train_batch_size = 2 * bsize * dist.get_world_size()
+
+    test_cfg.max_duration = '4ba'
+    test_cfg.eval_interval = '4ba'
+    test_cfg.eval_subset_num_batches = 2
+    test_cfg.save_interval = '4ba'
+    test_cfg.run_name = 'gpt-moe-test'
+    test_cfg.max_seq_len = 256
+
+    test_cfg.tokenizer.args.max_seq_len = test_cfg.max_seq_len
+    test_cfg.train_loader.dataset.max_seq_len = test_cfg.max_seq_len
+    test_cfg.eval_loader.dataset.max_seq_len = test_cfg.max_seq_len
+
+    test_cfg.model.max_seq_len = test_cfg.max_seq_len
+    test_cfg.model.d_model = 32
+    test_cfg.model.n_heads = 2
+    test_cfg.model.n_layers = 2
+    test_cfg.model.moe.num_experts = [2, 16]
+
+    return test_cfg
+
+
+def check_tensor(tensor):
+    world_size = dist.get_world_size()
+    tensors = dist.all_gather(tensor)
+    for i in range(world_size):
+        for j in range(i + 1, world_size):
+            assert not tensors[i].equal(tensors[j])
+            if tensors[i].all_close(tensors[j]):
+                warnings.warn(f'')
```

Comment:
I need to add an actual warning here.
If parameters are sync'd then `assert not tensors[i].equal(tensors[j])` will catch this as an issue.
If parameters are not sync'd but gradients are then eventually parameters will begin to be the similar over time. Which is why I make it into a warning; in general they're allowed to be close, but it might be indicative of an issue.

Reply:
rm